File: js/moq/src/ietf/publish_namespace.ts
import type * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Message from "./message.ts";
import * as Namespace from "./namespace.ts";
import { Parameters } from "./parameters.ts";

// In draft-14, ANNOUNCE is renamed to PUBLISH_NAMESPACE
export class PublishNamespace {
	static id = 0x06;

	requestId: number;
	trackNamespace: Path.Valid;

	constructor(requestId: number, trackNamespace: Path.Valid) {
		this.requestId = requestId;
		this.trackNamespace = trackNamespace;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
		await Namespace.encode(w, this.trackNamespace);
		await w.u8(0); // number of parameters
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishNamespace> {
		return Message.decode(r, PublishNamespace.#decode);
	}

	static async #decode(r: Reader): Promise<PublishNamespace> {
		const requestId = await r.u53();
		const trackNamespace = await Namespace.decode(r);
		await Parameters.decode(r); // ignore parameters
		return new PublishNamespace(requestId, trackNamespace);
	}
}

export class PublishNamespaceOk {
	static id = 0x07;

	requestId: number;

	constructor(requestId: number) {
		this.requestId = requestId;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishNamespaceOk> {
		return Message.decode(r, PublishNamespaceOk.#decode);
	}

	static async #decode(r: Reader): Promise<PublishNamespaceOk> {
		const requestId = await r.u53();
		return new PublishNamespaceOk(requestId);
	}
}

export class PublishNamespaceError {
	static id = 0x08;

	requestId: number;
	errorCode: number;
	reasonPhrase: string;

	constructor(requestId: number, errorCode: number, reasonPhrase: string) {
		this.requestId = requestId;
		this.errorCode = errorCode;
		this.reasonPhrase = reasonPhrase;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
		await w.u62(BigInt(this.errorCode));
		await w.string(this.reasonPhrase);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishNamespaceError> {
		return Message.decode(r, PublishNamespaceError.#decode);
	}

	static async #decode(r: Reader): Promise<PublishNamespaceError> {
		const requestId = await r.u53();
		const errorCode = Number(await r.u62());
		const reasonPhrase = await r.string();
		return new PublishNamespaceError(requestId, errorCode, reasonPhrase);
	}
}

export class PublishNamespaceCancel {
	static id = 0x0c;

	trackNamespace: Path.Valid;
	errorCode: number;
	reasonPhrase: string;

	constructor(trackNamespace: Path.Valid, errorCode: number = 0, reasonPhrase: string = "") {
		this.trackNamespace = trackNamespace;
		this.errorCode = errorCode;
		this.reasonPhrase = reasonPhrase;
	}

	async #encode(w: Writer): Promise<void> {
		await Namespace.encode(w, this.trackNamespace);
		await w.u62(BigInt(this.errorCode));
		await w.string(this.reasonPhrase);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishNamespaceCancel> {
		return Message.decode(r, PublishNamespaceCancel.#decode);
	}

	static async #decode(r: Reader): Promise<PublishNamespaceCancel> {
		const trackNamespace = await Namespace.decode(r);
		const errorCode = Number(await r.u62());
		const reasonPhrase = await r.string();
		return new PublishNamespaceCancel(trackNamespace, errorCode, reasonPhrase);
	}
}

// In draft-14, UNANNOUNCE is renamed to PUBLISH_NAMESPACE_DONE
export class PublishNamespaceDone {
	static readonly id = 0x09;

	trackNamespace: Path.Valid;

	constructor(trackNamespace: Path.Valid) {
		this.trackNamespace = trackNamespace;
	}

	async #encode(w: Writer): Promise<void> {
		await Namespace.encode(w, this.trackNamespace);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<PublishNamespaceDone> {
		return Message.decode(r, PublishNamespaceDone.#decode);
	}

	static async #decode(r: Reader): Promise<PublishNamespaceDone> {
		const trackNamespace = await Namespace.decode(r);
		return new PublishNamespaceDone(trackNamespace);
	}
}

File: rs/moq/src/ietf/publish_namespace.rs
//! IETF moq-transport-14 publish namespace messages

use std::borrow::Cow;

use crate::{coding::*, ietf::Message, Path};

use super::namespace::{decode_namespace, encode_namespace};

/// PublishNamespace message (0x06)
/// Sent by the publisher to announce the availability of a namespace.
#[derive(Clone, Debug)]
pub struct PublishNamespace<'a> {
	pub request_id: u64,
	pub track_namespace: Path<'a>,
}

impl<'a> Message for PublishNamespace<'a> {
	const ID: u64 = 0x06;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
		encode_namespace(w, &self.track_namespace);
		0u8.encode(w); // number of parameters
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		let track_namespace = decode_namespace(r)?;

		// Ignore parameters, who cares.
		let _params = Parameters::decode(r)?;

		Ok(Self {
			request_id,
			track_namespace,
		})
	}
}

/// PublishNamespaceOk message (0x07)
#[derive(Clone, Debug)]
pub struct PublishNamespaceOk {
	pub request_id: u64,
}

impl Message for PublishNamespaceOk {
	const ID: u64 = 0x07;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		Ok(Self { request_id })
	}
}

/// PublishNamespaceError message (0x08)
#[derive(Clone, Debug)]
pub struct PublishNamespaceError<'a> {
	pub request_id: u64,
	pub error_code: u64,
	pub reason_phrase: Cow<'a, str>,
}

impl<'a> Message for PublishNamespaceError<'a> {
	const ID: u64 = 0x08;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
		self.error_code.encode(w);
		self.reason_phrase.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		let error_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;

		Ok(Self {
			request_id,
			error_code,
			reason_phrase,
		})
	}
}
/// PublishNamespaceDone message (0x09)
#[derive(Clone, Debug)]
pub struct PublishNamespaceDone<'a> {
	pub track_namespace: Path<'a>,
}

impl<'a> Message for PublishNamespaceDone<'a> {
	const ID: u64 = 0x09;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.track_namespace);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let track_namespace = decode_namespace(r)?;
		Ok(Self { track_namespace })
	}
}

/// PublishNamespaceCancel message (0x0c)
#[derive(Clone, Debug)]
pub struct PublishNamespaceCancel<'a> {
	pub track_namespace: Path<'a>,
	pub error_code: u64,
	pub reason_phrase: Cow<'a, str>,
}

impl<'a> Message for PublishNamespaceCancel<'a> {
	const ID: u64 = 0x0c;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.track_namespace);
		self.error_code.encode(w);
		self.reason_phrase.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let track_namespace = decode_namespace(r)?;
		let error_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;
		Ok(Self {
			track_namespace,
			error_code,
			reason_phrase,
		})
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use bytes::BytesMut;

	fn encode_message<M: Message>(msg: &M) -> Vec<u8> {
		let mut buf = BytesMut::new();
		msg.encode(&mut buf);
		buf.to_vec()
	}

	fn decode_message<M: Message>(bytes: &[u8]) -> Result<M, DecodeError> {
		let mut buf = bytes::Bytes::from(bytes.to_vec());
		M::decode(&mut buf)
	}

	#[test]
	fn test_announce_round_trip() {
		let msg = PublishNamespace {
			request_id: 1,
			track_namespace: Path::new("test/broadcast"),
		};

		let encoded = encode_message(&msg);
		let decoded: PublishNamespace = decode_message(&encoded).unwrap();

		assert_eq!(decoded.track_namespace.as_str(), "test/broadcast");
	}

	#[test]
	fn test_announce_error() {
		let msg = PublishNamespaceError {
			request_id: 1,
			error_code: 404,
			reason_phrase: "Unauthorized".into(),
		};

		let encoded = encode_message(&msg);
		let decoded: PublishNamespaceError = decode_message(&encoded).unwrap();

		assert_eq!(decoded.error_code, 404);
		assert_eq!(decoded.reason_phrase, "Unauthorized");
	}

	#[test]
	fn test_unannounce() {
		let msg = PublishNamespaceDone {
			track_namespace: Path::new("old/stream"),
		};

		let encoded = encode_message(&msg);
		let decoded: PublishNamespaceDone = decode_message(&encoded).unwrap();

		assert_eq!(decoded.track_namespace.as_str(), "old/stream");
	}

	#[test]
	fn test_announce_cancel() {
		let msg = PublishNamespaceCancel {
			track_namespace: Path::new("canceled"),
			error_code: 1,
			reason_phrase: "Shutdown".into(),
		};

		let encoded = encode_message(&msg);
		let decoded: PublishNamespaceCancel = decode_message(&encoded).unwrap();

		assert_eq!(decoded.track_namespace.as_str(), "canceled");
		assert_eq!(decoded.error_code, 1);
		assert_eq!(decoded.reason_phrase, "Shutdown");
	}

	#[test]
	fn test_announce_rejects_parameters() {
		#[rustfmt::skip]
		let invalid_bytes = vec![
			0x01, // namespace length
			0x04, 0x74, 0x65, 0x73, 0x74, // "test"
			0x01, // INVALID: num_params = 1
		];

		let result: Result<PublishNamespace, _> = decode_message(&invalid_bytes);
		assert!(result.is_err());
	}
}
